[ClangImporter] Always load a class's members before any categories.

jrose-apple opened the pull request:
This ensures that any `@property` redeclarations that appear in class extensions (a special kind of category in ObjC) do not affect the primary type of the property as declared in the class itself. This is a better answer to #8324 today because it means the public and private cases look the same.

To accomplish this, lookups in ImportDecl.cpp that are checking for conflicts now no longer pull in new categories/extensions if the current context is a ClassDecl.

rdar://problem/30785976

jrose-apple committed:
[ClangImporter] Add PrettyStackTrace entries for import-as-member.

No functionality change.

jrose-apple committed:
[ClangImporter] Always load a class's members before any categories.

This ensures that any @property redeclarations that appear in class
extensions (a special kind of category in ObjC) do not affect the
primary type of the property as declared in the class itself.

To accomplish this, lookups in importDecl that are checking for
conflicts now no longer pull in new categories/extensions if the
current context is a ClassDecl.

rdar://problem/30785976

jrose-apple commented:
I'm much happier with this one but I definitely want Doug to look it over before I commit.

@swift-ci Please test

DougGregor commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4323,25 +4325,36 @@ namespace {
+        // If we're importing into the primary @interface for something, as
+        // opposed to an extension, make sure we don't try to load any
+        // categories...by just looking into the super type.
+        containerTy = classDecl->getSuperclass();
```
Can we rename `containerTy` to something more meaningful, now that it doesn't mean "the type into which this declaration will go". How about `lookupContainerTy`?

We don't need to look into the protocols to which this class conforms, because it's not an "override" in that case, right?

jrose-apple replied:
Hm. I don't actually know about protocols, but we don't *currently* look at them. It probably helps that we mirror all their members down for imported Clang things.

DougGregor commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -7495,6 +7510,15 @@ ClangImporter::Implementation::loadAllMembers(Decl *D, uint64_t extra) {
+      traceName = "(bridging header)";
+    PrettyStackTraceStringAction trace("loading import-as-members from",
+                                       traceName);
+    PrettyStackTraceDecl trace2("...for", nominal);
```
Trace greatly appreciated!

jrose-apple replied:
Eventually I will have a PrettyStackTrace for every operation in the compiler.

DougGregor reviewed (approved):
Yeah, this seems to be the best we can do, and it's certainly a safe, defensible rule.

jrose-apple committed:
[ClangImporter] Fix property redeclarations in generic classes.

Previously we were checking if a declaration was a redeclaration by
looking at the contextual type, but it should be looking at the
interface type in order to handle categories of classes using
Objective-C lightweight generics.

This also fixes a mistaken reuse of 'containerTy' after I changed its
meaning in the last commit. Doug's review comment was right: it /is/
bad to use an existing name for a new purpose!

More rdar://problem/30785976.

jrose-apple commented:
@swift-ci Please test
